src/stepFunctions/commands/visualizeStateMachine.ts

```diff
@@ -29,257 +29,346 @@ export interface messageObject {
+                return
+            }
+
+            logger.debug('Sending update message to webview.')
+
+            webview.postMessage({
+                command: 'update',
+                stateMachineData: updatedTextDocument.getText(),
+                isValid,
+            })
         }
-    })
-
-    // When the panel is closed, dispose of any disposables/remove subscriptions
-    panel.onDidDispose(() => {
-        updateOnSaveDisposable.dispose()
-        onDidCloseTextDocumentDisposable.dispose()
-        updateOnChangeDisposable.dispose()
-        receiveMessageDisposable.dispose()
-        debouncedUpdate.cancel()
-        isPanelDisposed = true
-    })
-
-    return panel
-}
+        const debouncedUpdate = debounce(sendUpdateMessage, 500).bind(this)
 
-function trackedDocumentDoesExist(trackedDocumentURI: vscode.Uri): boolean {
-    const document = vscode.workspace.textDocuments.find(doc => doc.fileName === trackedDocumentURI.fsPath)
+        this.disposables.push(
+            vscode.workspace.onDidChangeTextDocument(async textDocumentEvent => {
+                if (textDocumentEvent.document.uri.path === documentUri.path) {
+                    await debouncedUpdate(textDocumentEvent.document)
+                }
+            })
+        )
 
-    return document !== undefined
-}
+        // Handle messages from the webview
+        this.disposables.push(
+            panel.webview.onDidReceiveMessage(async (message: messageObject) => {
+                switch (message.command) {
+                    case 'updateResult':
+                        logger.debug(message.text)
+                        if (message.error) {
+                            logger.error(message.error)
+                        }
+                        break
+                    case 'webviewRendered': {
+                        // Webview has finished rendering, so now we can give it our
+                        // initial state machine definition.
+                        await sendUpdateMessage(textDocument)
+                        break
+                    }
+
+                    case 'viewDocument':
+                        try {
+                            const document = await vscode.workspace.openTextDocument(documentUri)
+                            vscode.window.showTextDocument(document, vscode.ViewColumn.One)
+                        } catch (e) {
+                            logger.error(e as Error)
+                        }
+                        break
+                }
+            })
+        )
 
-function makeWebviewTitle(sourceDocumentUri: vscode.Uri): string {
-    return localize('AWS.stepFunctions.graph.titlePrefix', 'Graph: {0}', path.basename(sourceDocumentUri.fsPath))
-}
+        // When the panel is closed, dispose of any disposables/remove subscriptions
+        panel.onDidDispose(() => {
+            this.onVisualizationDisposeEmitter.fire()
+            for (const disposable of this.disposables) {
```

Comment:
nit. I think forEach would be more readable here:  `this.disposables.forEach(disposable => disposable.dispose())`